src/objects.h

```diff
@@ -66,6 +66,20 @@ class param_type
   std::vector<std::size_t> vec_size_t_;
 };
 
+template <> inline double & param_type::data<double>() { return double_; }
+template <> inline int & param_type::data<int>() { return int_; }
+template <> inline bool & param_type::data<bool>() { return bool_; }
+template <> inline std::vector<double> & param_type::data<std::vector<double>>() { return vec_double_; }
+template <> inline std::shared_ptr<NEMLObject> & param_type::data<std::shared_ptr<NEMLObject>>() { return neml_object_; }
+template <> inline std::vector<std::shared_ptr<NEMLObject>> & param_type::data<std::vector<std::shared_ptr<NEMLObject>>>() { return vec_neml_object_; }
+template <> inline std::string & param_type::data<std::string>() { return string_; }
+template <> inline list_systems & param_type::data<list_systems>() { return list_systems_; }
+template <> inline std::size_t & param_type::data<std::size_t>() { return size_t_; }
+template <> inline std::vector<std::size_t> & param_type::data<std::vector<std::size_t>>() { return vec_size_t_; }
+
+inline param_type::param_type(const char * val) { data<std::string>() = *val; }
```

Comment:
could you add a comment about the last inline? it's confusing because it does not return anything but sotres things inside another variable

Reply:
It converts the const char* type to a string.  I had to move these all to be `inline` because apparently Daniel provided us with code that only compiles on g++ (or at least doesn't compile on mingw...).